Refactor ref src generators & pass ref assemblies

ViktorHofer opened the pull request:
Fixes https://github.com/dotnet/source-build/issues/3252
Fixes https://github.com/dotnet/source-build/issues/3237

- Refactor textonly & reference source generators & apply general clean-up
- Pass the right reference assemblies to GenAPI invocation
- Make GenAPI generate assembly attributes and typeforwards
- Use PackageDownload instead of extra restore project
- Use the default package cache instead of defining a separate one
- Make the ref generator work with multi-assembly packages
- Make the generators work cross-plat
- Add "--exclude-package-dependencies" switch to ignore dependencies

cc @andriipatsula @MichaelSimons

andriipatsula commented:
Hello @ViktorHofer , I tried to run the generator from a PR branch and got an error:  "The "MSBuildInParallel" parameter is not supported by the "MSBuild" task loaded from assembly". And I took the latest commit.

>  Determining projects to restore...
  Restored /root/.nuget/packages/microsoft.dotnet.arcade.sdk/8.0.0-beta.23120.1/tools/Tools.proj (in 18.59 sec).
  Determining projects to restore...
  Restored /repos/source-build-reference-packages/src/referencePackageSourceGenerator/ReferencePackageSourceGenerator.csproj (in 25.82 sec).
  ReferencePackageSourceGenerator -> /repos/source-build-reference-packages/artifacts/bin/ReferencePackageSourceGenerator/Debug/net8.0/ReferencePackageSourceGenerator.dll
/repos/source-build-reference-packages/src/referencePackageSourceGenerator/ReferencePackageSourceGenerator.csproj(37,14): **error MSB4064: The "MSBuildInParallel" parameter is not supported by the "MSBuild" task loaded from assembly: Microsoft.Build, Version=15.1.0.0**, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a from the path: /repos/source-build-reference-packages/.dotnet/sdk/8.0.100-alpha.1.23061.8/Microsoft.Build.dll. **Verify that the parameter exists on the task, the <UsingTask> points to the correct assembly, and it is a settable public instance property.**
/repos/source-build-reference-packages/src/referencePackageSourceGenerator/ReferencePackageSourceGenerator.csproj(31,5): **error MSB4063: The "MSBuild" task could not be initialized with its input parameters**

MichaelSimons commented:
``` bash
> ./generate.sh --pkg System.Threading.Tasks,4.3.0
  Determining projects to restore...
  All projects are up-to-date for restore.
/repos/source-build-reference-packages/src/referencePackageSourceGenerator/ReferencePackageSourceGenerator.csproj(9,5): error MSB4184: The expression "[MSBuild]::NormalizeDirectory('', '')" cannot be evaluated. Parameter "path" cannot have zero length.
```

ViktorHofer commented:
Taking a look

andriipatsula commented:
One more thing @MichaelSimons ,
I compare results of the SBRP generator from the `main` branch and from the PR. Package `microsoft.codeanalysis.common,4.4.0`

The version from main generates packages and direct dependencies and the version from the PR - direct and transitive dependencies.

What 'd be a desired behavior? The same as original version of SBRP?

## From the main:
**microsoft.codeanalysis.common
system.buffers
system.collections.immutable
system.memory
system.numerics.vectors
system.reflection.metadata
system.runtime.compilerservices.unsafe
system.text.encoding.codepages
system.threading.tasks.extensions**

## From the PR branch:
**microsoft.codeanalysis.common
system.buffers
system.collections.immutable
system.memory
system.numerics.vectors
system.reflection.metadata
system.runtime.compilerservices.unsafe
system.text.encoding.codepages
system.threading.tasks.extensions**
microsoft.codeanalysis.analyzers
microsoft.netcore.platforms
microsoft.netcore.targets
microsoft.win32.primitives
netstandard.library
system.appcontext
system.collections
system.collections.concurrent
system.console
system.diagnostics.debug
system.diagnostics.diagnosticsource
system.diagnostics.tools
system.diagnostics.tracing
system.globalization
system.globalization.calendars
system.globalization.extensions
system.io
system.io.compression
system.io.compression.zipfile
system.io.filesystem
system.io.filesystem.primitives
system.linq
system.linq.expressions
system.net.http
system.net.primitives
system.net.sockets
system.objectmodel
system.reflection
system.reflection.emit
system.reflection.emit.ilgeneration
system.reflection.emit.lightweight
system.reflection.extensions
system.reflection.primitives
system.reflection.typeextensions
system.resources.resourcemanager
system.runtime
system.runtime.extensions
system.runtime.handles
system.runtime.interopservices
system.runtime.interopservices.runtimeinformation
system.runtime.numerics
system.security.cryptography.algorithms
system.security.cryptography.cng
system.security.cryptography.csp
system.security.cryptography.encoding
system.security.cryptography.openssl
system.security.cryptography.primitives
system.security.cryptography.x509certificates
system.text.encoding
system.text.encoding.extensions
system.text.regularexpressions
system.threading
system.threading.tasks
system.threading.timer
system.xml.readerwriter
system.xml.xdocument

ViktorHofer commented:
@MichaelSimons the error above is now fixed. The generate.sh script always appended msbuild properties even if they weren't set which led to the default path not being used and being empty. I changed that to only append the switches when they are passed-in.

ViktorHofer commented:
With the current branch, the following dependency closure gets transformed to reference package source when invoking `generate.sh --package microsoft.codeanalysis.common,4.4.0`

Microsoft.CodeAnalysis.Common (v4.4.0)
Microsoft.CodeAnalysis.Analyzers (v3.3.3)
System.Collections.Immutable (v6.0.0)
System.Runtime.CompilerServices.Unsafe (v6.0.0)
System.Memory (v4.5.4)
System.Buffers (v4.5.1)
System.Diagnostics.Debug (v4.3.0)
Microsoft.NETCore.Platforms (v1.1.0)
Microsoft.NETCore.Targets (v1.1.0)
System.Runtime (v4.3.0)
System.Diagnostics.Tracing (v4.3.0)
System.Resources.ResourceManager (v4.3.0)
System.Globalization (v4.3.0)
System.Reflection (v4.3.0)
System.IO (v4.3.0)
System.Text.Encoding (v4.3.0)
System.Threading.Tasks (v4.3.0)
System.Reflection.Primitives (v4.3.0)
System.Threading (v4.3.0)
System.Runtime.CompilerServices.Unsafe (v4.5.3)
System.Numerics.Vectors (v4.5.0)
System.Runtime.Extensions (v4.3.0)
System.Runtime.InteropServices (v4.3.0)
System.Runtime.Handles (v4.3.0)
System.Numerics.Vectors (v4.4.0)
System.Memory (v4.5.5)
System.Reflection.Metadata (v5.0.0)
System.Collections.Immutable (v5.0.0)
System.Text.Encoding.CodePages (v6.0.0)
System.Threading.Tasks.Extensions (v4.5.4)
System.Collections (v4.3.0)

@andriipatsula's previous list was longer because we didn't filter out targeting packs before. That's now fixed. The remaining issue is that text only packages are considered during the reference package source traversal:

Microsoft.CodeAnalysis.Analyzers (v3.3.3)
Microsoft.NETCore.Platforms (v1.1.0)
Microsoft.NETCore.Targets (v1.1.0)

We could either ignore them (existing behavior) or pass them to the text only package generator. Alternatively, we could also merge both generators so that it can handle both types of packages.

MichaelSimons commented:
My main requirement is that text-only packages remain in a separate location of the repo.  It is important that their is a distinction made between them and ref packages because the rules for consuming them are very different.  There is a bar for adding new ones (e.g. it would be preferred that we truly source-build them). Because of that I do think it is preferable that you must take an explicit action to generate them.

I see value in having a single set of tools that would support either.  My only request would be to do any additional refactoring separately as we would like to realize the immediate benefit of this PR.

_What was the reason that made us not include allowed content in reference source packages?_ - I don't have the history here.  I can speculate one reason was to avoid bloat from non-required content for the required usage scenarios.  We are trying to reduce the size of this repo so bringing in a bunch of things like icons would work against this for no real value.

Regarding text-only packages referenced by reference packages, I think we should continue to ignore them as is the current behavior.  We can reconsider this if it becomes a UX concern.

ViktorHofer commented:
The PR should now be ready to be merged. I tested it with Microsoft.CodeAnalysis.Common + its dependencies and with System.Drawing.Common + dependency.

- Text only packages are now skipped (with a message) but the code is already written so that we can easily trigger the text only package generator in the future (when such a package appears as a dependency), if we would want that.
- I also noticed that the project file had the wrong casing and the targeting pack dependencies were missing. That's now fixed.
- The generated Directory.Build.props file receives the PackageId as an input which is incorrect as it should get the assembly name instead. In most cases those are in sync but not in all. That said, this bug already exists in the current tooling in main.

ViktorHofer commented:
@MichaelSimons would you mind taking another look and if everything's good, merge the PR in?

MichaelSimons commented on eng/Versions.props:
```diff
@@ -11,11 +11,15 @@
     <MicrosoftNETCoreILDAsmVersion>6.0.0-preview.6.21352.12</MicrosoftNETCoreILDAsmVersion>
     <!-- SDK dependencies -->
     <MicrosoftDotNetGenAPITaskPackageVersion>8.0.100-preview.3.23152.1</MicrosoftDotNetGenAPITaskPackageVersion>
+    <!-- NuGet dependencies -->
```
Consider alphabetizing dependency groupings.

MichaelSimons commented:
How should I interpret the following errors when running `./generate.sh -pkg microsoft.build,16.5.0`?   I feel like this is something we need guidance on.

/root/.nuget/packages/microsoft.dotnet.genapi.task/8.0.100-preview.3.23152.1/build/Microsoft.DotNet.GenAPI.Task.targets(44,5): error CP1002: AssemblyLoadWarning: 'Could not resolve reference 'System.ServiceModel.Internals.dll' in any of the provided search directories.' [/repos/source-build-reference-packages/artifacts/bin/ReferencePackageSourceGenerator/microsoft.build/16.5.0/microsoft.build.16.5.0.csproj::TargetFramework=net472]
/root/.nuget/packages/microsoft.dotnet.genapi.task/8.0.100-preview.3.23152.1/build/Microsoft.DotNet.GenAPI.Task.targets(44,5): error CP1002: AssemblyLoadWarning: 'Could not resolve reference 'SMDiagnostics.dll' in any of the provided search directories.' [/repos/source-build-reference-packages/artifacts/bin/ReferencePackageSourceGenerator/microsoft.build/16.5.0/microsoft.build.16.5.0.csproj::TargetFramework=net472]
/root/.nuget/packages/microsoft.dotnet.genapi.task/8.0.100-preview.3.23152.1/build/Microsoft.DotNet.GenAPI.Task.targets(44,5): error CP1002: AssemblyLoadWarning: 'Could not resolve reference 'System.ServiceModel.Internals.dll' in any of the provided search directories.' [/repos/source-build-reference-packages/artifacts/bin/ReferencePackageSourceGenerator/system.runtime.serialization.primitives/4.1.1/system.runtime.serialization.primitives.4.1.1.csproj::TargetFramework=net46]
/root/.nuget/packages/microsoft.dotnet.genapi.task/8.0.100-preview.3.23152.1/build/Microsoft.DotNet.GenAPI.Task.targets(44,5): error CP1002: AssemblyLoadWarning: 'Could not resolve reference 'SMDiagnostics.dll' in any of the provided search directories.' [/repos/source-build-reference-packages/artifacts/bin/ReferencePackageSourceGenerator/system.runtime.serialization.primitives/4.1.1/system.runtime.serialization.primitives.4.1.1.csproj::TargetFramework=net46]
/repos/source-build-reference-packages/artifacts/bin/ReferencePackageSourceGenerator/system.runtime/4.1.0/system.runtime.4.1.0.csproj : error NU1605: Warning As Error: Detected package downgrade: System.Runtime from 4.3.0 to 4.1.0. Reference the package directly from the project to select a different version. 
/repos/source-build-reference-packages/artifacts/bin/ReferencePackageSourceGenerator/system.runtime/4.1.0/system.runtime.4.1.0.csproj : error NU1605:  system.runtime.4.1.0 -> NETStandard.Library 1.6.1 -> System.Runtime (>= 4.3.0) 
/repos/source-build-reference-packages/artifacts/bin/ReferencePackageSourceGenerator/system.runtime/4.1.0/system.runtime.4.1.0.csproj : error NU1605:  system.runtime.4.1.0 -> System.Runtime (>= 4.1.0)
    0 Warning(s)
    5 Error(s)

MichaelSimons commented:
Another scenario:  `./generate.sh -pkg microsoft.codeanalysis.csharp,4.0.1`

```
/repos/source-build-reference-packages/.dotnet/sdk/8.0.100-alpha.1.23061.8/NuGet.targets(132,5): error : '[4.0.1]' is not a valid version string. (Parameter 'value') [/repos/source-build-reference-packages/src/referencePackageSourceGenerator/ReferencePackageSourceGenerator.proj]
```

In this scenario, only the microsoft.codeanalysis.csharp,4.0.1 package way generated.  It's dependency graph was not.

ViktorHofer commented on src/referencePackageSourceGenerator/ReferencePackageSourceTask/GenerateProject.cs:
```diff
@@ -0,0 +1,175 @@
+        [Required]
+        public string ProjectTemplate { get; set; }
+
+        public override bool Execute()
```
I left it as is as this was already a single method in main: https://github.com/dotnet/source-build-reference-packages/blob/a3ee2277e860df151d33d72e9b7fabf1e564b3d8/src/referencePackageSourceGenerator/Tasks/GenerateProjects.cs#L32-L225

While I agree with the principle of factoring responsibilities into different methods, I would recommend to follow-up on that in a separate change. Also, this PR already adds ton of code comments and significantly reduces the code size because of simplification.

MichaelSimons replied:
That's fair, with all the refactoring I lost sight of original code.  I appreciate the effort you invested to improve the codebase - thanks.

ViktorHofer commented:
> How should I interpret the following errors when running ./generate.sh -pkg microsoft.build,16.5.0? I feel like this is something we need guidance on.

The ´AssemblyLoadWarning` warnings are reasonable and should stay. In the case of .NET Framework, reference assemblies are missing (intentionally AFAIK as those never got exposed in the reference assemblies pack). Those warnings were propagated to errors which I just fixed.

The package downgrade is intentional, as we are referencing a specific version of a package which might be a lower version than what other dependencies (targeting pack) might bring in. I just added a NoWarn for it.

> In this scenario, only the microsoft.codeanalysis.csharp,4.0.1 package way generated. It's dependency graph was not.

Fixed. That package had fixed versions for its dependencies and the tooling / NuGet APIs handled that incorrectly.
